### PR TITLE
Exposes GOES quality flags

### LINKED ISSUE #6259: Expose GOES quality flags
### Describe the feature

Currently, it is not possible to differentiate between corrupt and normal values when using the GOES `XRSTimeSeries`. As there are quite a few corrupt values which can be seen in the following image (see number of X-class values), it would be great to expose the flags indicating whether a given value is corrupt or not as well.
173183771-b030c0e6-6897-4752-a341-e17b3e054658
(also refer to https://satdat.ngdc.noaa.gov/sem/goes/data/science/xrs/GOES_13-15_XRS_Science-Quality_Data_Readme.pdf)

### Proposed solution

Exposes GOES quality flags (xrsa_quality, xrsb_quality) as part of the `XRSTimeSeries` in order to filter for invalid values. An example could look as follows:

```
from sunpy import timeseries as ts
from sunpy.net import Fido
from sunpy.net import attrs as a

t_start = "2014-01-01 00:00"
t_end = "2014-01-31 00:00"
search_result = Fido.search(
                        a.Time(t_start, t_end), a.Instrument.xrs)
download_result = Fido.fetch(search_result)
goes_ts = ts.TimeSeries(download_result, concatenate=True)
df = goes_ts.to_dataframe()
df_clean = df[(df["xrsb_quality"] == 0)]
```

A PR will follow

### PR DESCRIPTION
**Pull request by mariusgiger**
## PR Description

Resolves #6259 by exposing GOES quality flags to the user

TODO how to extract quality flags from the FITS header?

### CODE REVIEW

File: sunpy/timeseries/sources/goes.py
Comment (mariusgiger) [inline sunpy/timeseries/sources/goes.py:299]: Is it better to name this xrsa_flags as well?

File: sunpy/timeseries/sources/goes.py
Comment (nabobalis) [inline sunpy/timeseries/sources/goes.py:221]: Do you have a FITS file to test with? I would hope it's one of the array rows in a HDU.

File: sunpy/timeseries/sources/goes.py
Comment (nabobalis) [inline sunpy/timeseries/sources/goes.py:281]: Are there other columns that would be useful to have extracted too?

File: sunpy/timeseries/sources/goes.py
Comment (mariusgiger) [reply sunpy/timeseries/sources/goes.py:221]: When looking at the FITS header the following information is contained,
e.g. for https://umbra.nascom.nasa.gov/goes/fits/2019/go1520191210.fits:

```
import sunpy.io
from sunpy.util.metadata import MetaDict
from collections import OrderedDict

filepath= "./go1520191210.fits"
hdulist = sunpy.io.read_file(filepath)
header = MetaDict(OrderedDict(hdulist[3].header))
print(header)
print(hdulist[3].data)
>FITS_rec([(0.55999994, [0., 0.])],
         dtype=(numpy.record, [('Time', '>f4'), ('Status', '>f4', (2,))]))
```

The following doc is provided for `Status`, however it is not clear to me how the status maps to xrsa and xrsb (e.g. in netcdf format there is a value for each channel and measurement):

Convert status words to long integer then examine bits.Status 
1 - octal mask, description (incomplete list):
- 1000, Sun eclipsed by MoonStatus 
2 - octal mask, description (incomplete list):
- 1, X-Ray detector off
- 2, X-Ray detector being calibrated
- 4, X-Ray Transient
- 10, X-Ray short channel saturation
- 20, X-Ray long channel range change
- 40, X-Ray short channel range change
- 200, X-Ray long channel saturation')

Does someone know how to extract this information?

File: sunpy/timeseries/sources/goes.py
Comment (mariusgiger) [reply sunpy/timeseries/sources/goes.py:281]: other variables in netcdf - not sure if any of those should also be exposed:

```
xrsa1_flux: ('time',) float32
xrsa2_flux: ('time',) float32
xrsb1_flux: ('time',) float32
xrsb2_flux: ('time',) float32
xrsa_primary_chan: ('time',) uint8
xrsb_primary_chan: ('time',) uint8
xrsa1_flags: ('time',) uint16
xrsa2_flags: ('time',) uint16
xrsb1_flags: ('time',) uint16
xrsb2_flags: ('time',) uint16
corrected_current_xrsa2: ('time', 'quad_diode') float32
corrected_current_xrsb2: ('time', 'quad_diode') float32
au_factor: ('time',) float32
roll_angle: ('time',) float32
yaw_flip_flag: ('time',) uint8
```

File: sunpy/timeseries/sources/goes.py
Comment (mariusgiger) [reply sunpy/timeseries/sources/goes.py:299]: the value is a flag based on the following mask:

```
[0,   1,   2,   4,   8,  16,  32,  64, 128, 256, 512]
[good_data, eclipse, particle_spike, calibration, off_point, temperature_error, data_quality_error, pointing_error, invalid_mode, missing_data, L0_error]
```

any value different from 0 means something is not correct

e.g. 72 = 64 + 8 = pointing_error + off_point

**Comment by mariusgiger** [comment]
@wtbarnes yes this could also be an approach - the question is which values to mask without hiding something from the user that should actually be available.

One example could be:

```
ts['xrsb'].mask = ts['xrsb_flags'] == 0
```

**Comment by dstansby** [comment]
It's not possible to mask data in pandas as far as I know. I think we should limit the scope of this PR to returning the quality flags, and backporting as a bugfix. @mariusgiger could you add a short changelog entry in the `changelog` directory (there's instructions on what to do in there? Once that's done I'm happy to approve this.

**Comment by dstansby** [comment]
It's not possible to mask data in pandas as far as I know. I think we should limit the scope of this PR to returning the quality flags, and backporting as a bugfix. @mariusgiger could you add a short changelog entry in the `changelog` directory (there's instructions on what to do in there? Once that's done I'm happy to approve this.

File: sunpy/timeseries/sources/goes.py
Comment (dstansby) [reply sunpy/timeseries/sources/goes.py:221]: I think this is beyond scope of this PR - could someone open an issue so we can keep track of this though?

Comment (dstansby) [reply sunpy/timeseries/sources/goes.py:221]: I think this is beyond scope of this PR - could someone open an issue so we can keep track of this though?

File: sunpy/timeseries/sources/goes.py
Comment (dstansby) [reply sunpy/timeseries/sources/goes.py:281]: I think we can leave these, unless someone else requests one of them with a reason.

File: sunpy/timeseries/sources/goes.py
Comment (nabobalis) [reply sunpy/timeseries/sources/goes.py:221]: I would rather we didn't add support only for one file type but I don't want to block this PR on that.

File: examples/acquiring_data/goes_xrs_example.py
Comment (dstansby) [inline examples/acquiring_data/goes_xrs_example.py:76]:
```suggestion
goes_15 = ts.TimeSeries(df, goes_15.meta, goes_15.units)
```
This is the best way to do this - if we change the internals of how `TimeSeries` works in the future this will avoid relying `.data` being a dataframe.

**Review by dstansby — CHANGES_REQUESTED** [review]
Other than my single comment, looks 👍

**Review by dstansby — APPROVED** [review]

**Comment by nabobalis** [comment]
Thanks for the PR @mariusgiger, I have created a tracking issue about the FITS files.

**Pull request merged**
